src/Sound/SoundPack.cpp

```diff
@@ -0,0 +1,112 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2020 openblack developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/openblack/openblack
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#include "SoundPack.h"
+
+#include <Common/FileSystem.h>
+#include <Game.h>
+#include <spdlog/spdlog.h>
+
+
+namespace openblack::audio
+{
+
+void SoundPack::LoadFromFile(const std::string& filename)
+{
+	spdlog::debug("Loading Sound Pack from file: {}", filename);
+
+	try
+	{
+		_sadFile.Open(Game::instance()->GetFileSystem().FindPath(filename).u8string());
+	}
+	catch (std::runtime_error& err)
+	{
+		spdlog::error("Failed to open {}: {}", filename, err.what());
+		return;
+	}
+
+	if (!_sadFile.GetName().empty())
+		_name = _sadFile.GetName();
+	else
+		_name = (filename);
+
+	// TODO: Discover sector flag
+	bool isSector;
+	auto combinedData = std::map<std::string, Sound>();
+
+	for (auto& [sample, soundData] : _sadFile.GetSounds())
+	{
+		auto path = fs::path(sample.audioSampleName).filename();
+		auto soundName = path.string();
+		auto ext = path.extension().string();
+		isSector = ext == ".mpg";
+		auto volume = sample.volume > 0 ? ((float)sample.volume) / 127.f : 1.f;
+
+		// We'll use the file name for the sound ID instead as each sector has different name
+		if (isSector)
+		{
+			if (combinedData.find(filename) == combinedData.end())
+				combinedData[filename] = {
```

Comment:
During my tests, at this point, `filename` is the whole path of the file. This is likely why fetching the id of the sound fails in other parts. You should be using the basename of the path.

---

Also if readability is a concern do this.
```c++
auto emplacement = combinedData.emplace(filename, Sound{});
auto& sound = emplacement.first->second;
if (emplacement.second) // Sound was not in the map
{
	sound.id = filename;
	sound.name = filename;
	sound.priority = sample.priority;
	sound.sampleRate = static_cast<int>(sample.sampleRate);
	sound.bitRate = 0;
	sound.volume = volume;
	sound.pitch = sample.pitch;
	sound.pitchDeviation = sample.pitchDeviation;
	sound.bytes = soundData;
	sound.channelLayout = ChannelLayout::Stereo;
	sound.playType = static_cast<PlayType>(sample.loopType);
	sound.loaded = false;
	sound.sectorMerged = true;
	sound.sectorCount = 1;
	sound.lengthInSeconds = -1;
}
else // Sound was already in the map
{
	sound.sectorCount++;
	sound.bytes.insert(sound.bytes.end(), soundData.begin(), soundData.end());
}
```
to replace the whole if statement `if (combinedData.find(filename) == combinedData.end()) ... else`

Reply:
I prefer the suggested change with emplacement because already the comment for `sound.playType` was mislabled